Calculate errors on Discus results

DannyHindson opened the pull request:
**Description of work.**

Calculate the errors on the weights calculated by DiscusMultipleScatteringCorrection. Follow similar approach to MonteCarloAbsorption where the error is calculated based on the standard deviation of the weights per scenario divided by the square root of the number of scenarios. When interpolation is used combine the errors from the simulated points using standard error propagation formulae and combine with interpolation error when using linear interpolation.

As part of this I've fixed a few of issues:
- a crash in workbench when plotting error bars on a spectrum where any of the points has an error=NaN
- a bug in the the python export of MatrixWorkspace where the setE function only worked if a numpy array was supplied. The equivalent setY function had been previously modified to accept lists. It's helpful for them to accept lists because if you supply a flipped numpy array the setY\setE functions don't work (see https://github.com/mantidproject/mantid/issues/33673)
- the error calculation when using spline interpolation involves calling Matrix<T>::invertTridiagonal. This function has a check in it to see if a matrix was tridiagonal and the check wasn't handling the numerical precision properly. It used `std::numeric_limits<double>::epsilon()` but this needs scaling by the numbers being compared

**To test:**

Run the second usage example from the DiscusMultipleScatteringCorrection documentation (without the final mtd.clear() line).
Right click in the plot area and select Error Bars, All Errors
Observe that visible error bars appear for the double scatter lines. The errors on the single scatter lines are much smaller and aren't visible:
![image](https://user-images.githubusercontent.com/56295817/204315899-7f4f724d-f0c1-4b91-8950-a88af1b16dcb.png)
Change the NeutronPathsMultiple setting to 10,000 and rerun and observe that the errors get smaller


Fixes #34770.

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Fix crash when showing errors if workspace contains a nan error

DannyHindson committed:
Add error calculation

DannyHindson committed:
Allow setE to accept python lists as well as arrays

Same as setX and setY
This allows you to avoid the bug where a flipped array is supplied
and the C++ code mangles it by supplying a list instead - generated
via a.tolist()

DannyHindson committed:
Include errors when flipping data in usage example

DannyHindson committed:
Add comment about downside of spline interpolation

Noticed when using the interpolation feature on DiscusMultipleScatteringCorrection
on the w axis of a Lorentzian with sharp peak at w=0

DannyHindson committed:
Improve comments around spline interpolation method used

DannyHindson committed:
Improve tolerance check as to whether matrix is tridiagonal

Original use of std::numeric_limits<double>::epsilon() was invalid
because it needs scaling by the value(s) being compared
Change Matrix class to take a simple tolerance and then end user
can choose whether to use epsilon to calculate this or not

DannyHindson committed:
Add release note

DannyHindson committed:
Resolve shadow variable declaration problem

DannyHindson committed:
Log warning instead of stopping if not enough points for error calc

DannyHindson committed:
Add test to check errors calculated

DannyHindson committed:
Use FLT_EPSILON rather than Kernel::Tolerance

because already used in unit

DannyHindson committed:
Use Integration algorithm to obtain errors

Use Integration instead of local integrateCumulative function so that can
get errors in _integrated workspaces
Also add test to check errors are output on all workspaces

DannyHindson committed:
Adjust interpolation test now that it now longer throws

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -1035,9 +1042,9 @@ void DiscusMultipleScatteringCorrection::integrateCumulative(const DiscusData1D
                                                              const double xmax, std::vector<double> &resultX,
                                                              std::vector<double> &resultY,
                                                              const bool returnCumulative) {
+  assert(h.histogram().xMode() == HistogramData::Histogram::XMode::Points);
```
I'm having a problem building this - I get an error on this assert line
```
4>C:\mantid-conda\mantid\Framework\Algorithms\src\DiscusMultipleScatteringCorrection.cpp(1045,3): error C2039: 'histogram': is not a member of 'Mantid::Algorithms::DiscusData1D'
4>C:\mantid-conda\mantid\Framework\Algorithms\inc\MantidAlgorithms/DiscusMultipleScatteringCorrection.h(36): message : see declaration of 'Mantid::Algorithms::DiscusData1D'
```
`DiscusData1D` is a struct with x and y and no histogram method that I can see...
Perhaps it's slightly misleading for `DiscusData2D.hitogram()` to produce a `DiscusData1D` struct - or maybe the `DiscusData1D` needs an xmode attribute?
Or maybe I'm not building it right as the jenkins tests seem OK...

DannyHindson replied:
Yes this was a genuine problem. I checked and a local Release build worked with this in place so the compiler must disregard these lines when doing a Release build (which is what Jenkins does). I'll fix

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -1342,14 +1345,26 @@ std::vector<double> DiscusMultipleScatteringCorrection::simulatePaths(
       std::transform(weights.begin(), weights.end(), sumOfWeights.begin(), sumOfWeights.begin(), std::plus<double>());
       std::transform(weights.begin(), weights.end(), countZeroWeights.begin(), countZeroWeights.begin(),
                      [](double d, int count) { return d > 0. ? count : count + 1; });
```
Compiler also warns about precision loss here, I thought that might get caught by cppcheck?
```
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include\algorithm(2184,36): warning C4244: 'argument': conversion from 'double' to 'int', possible loss of data
1>C:\mantid-conda\mantid\Framework\Algorithms\src\DiscusMultipleScatteringCorrection.cpp(1347): message : see reference to function template instantiation '_OutIt std::transform<std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>,std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>,std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>,Mantid::Algorithms::DiscusMultipleScatteringCorrection::simulatePaths::<lambda_971e91370384df0d12d827aa27767530>>(const _InIt1,const _InIt1,const _InIt2,_OutIt,_Fn)' being compiled
```

DannyHindson replied:
I didn't see this compiler warning from Visual Studio but I can see why it might be a problem. I've fixed this by changing the type of the vector `countZeroWeights` to int

DannyHindson committed:
Fix compiler error in assert statement

RichardWaiteSTFC commented on Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:
```diff
@@ -1342,14 +1345,26 @@ std::vector<double> DiscusMultipleScatteringCorrection::simulatePaths(
+        weightsM2[i] += deltas[i] * (weights[i] - weightsMeans[i]);
+        // calculate sample SD (M2/n-1)
+        // will give NaN for m_events=1, but that's correct
+        weightsErrors[i] = sqrt(weightsM2[i] / static_cast<double>(ie));
```
I think the variance doesn't need to be divided by `ie` here (the mean is updated by the deviation weighted by `ie+1` and then the product of the deviation from the old mean and the deviation from the new mean is added to the variance sum.
At a later point you divide by `sqrt(N-1)` in another loop, so maybe it's a typo and really you should divide by `sqrt(nPaths)` inside this loop and ditch the later loop?

DannyHindson replied:
I've had a think about this. I think the division by `ie` on L1356 corresponds to the "M2 / count" here from the Wikipedia page that I've highlighted in yellow:
![image](https://user-images.githubusercontent.com/56295817/208728628-dd0f5630-95d8-463e-b074-e0500bf7e29d.png)
This sets weightsErrors[i] equal to the standard deviation of the weights for a particular omega value. I probably don't need to do L1356 on every iteration of the loop - each execution of L1356 just overwrites the previous answer but the repetition is harmless. **Note** - at this point weightsErrors[i] is just a standard deviation and not an error. The naming of the vector here isn't great because it's based on what the vector will hold at the end of the function.

I then need to do a further division by sqrt(nPaths) on L1364 to get the error on the mean weight for a particular omega value. This bit isn't on the Wikipedia page at all because the Welford algorithm is all about getting the variance\SD of a set of values. It doesn't cover getting the error on the mean. Does that make sense?

As discussed this is perhaps easier to see in MCAbsorptionStrategy.cpp. I tried removing the final division in MCAbsorptionStrategy.cpp and running the MonteCarloAbsorption algorithm using this script and the errors on the abscor workspace seemed too large which gives me a bit of confidence my argument above is valid:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

data = CreateSampleWorkspace(WorkspaceType='Histogram', NumBanks=1)
data = ConvertUnits(data, Target="Wavelength")
# Default up axis is Y
SetSample(data, Geometry={'Shape': 'Cylinder', 'Height': 5.0, 'Radius': 1.0,
                  'Center': [0.0,0.0,0.0]},
                Material={'ChemicalFormula': '(Li7)2-C-H4-N-Cl6', 'SampleNumberDensity': 0.07})
# Simulating every data point can be slow. Use a smaller set and interpolate
abscor = MonteCarloAbsorption(InputWorkspace=data,EventsPerPoint=1000)
corrected = data/abscor
```

RichardWaiteSTFC replied:
I think there is a bug, to show it I converted the code to python - so I may have made an error, but I think this is what the code is doing (converted into python)
```
import numpy as np

avg = 1
std = 0.2

nomega = 5
npaths = 1000
weights = np.random.normal(loc=avg, scale=std, size=(npaths, nomega))
# calculate the avg and std for each omega
print('avg = ', np.mean(weights, axis=0))
print('std = ', np.std(weights, axis=0))
# avg =  [0.9997892  0.99953238 0.99211105 1.01015046 1.00047634]
# std =  [0.19445069 0.20230386 0.20706431 0.20533679 0.19615291]

# use welford one-pass as in DiscusMultipleScatteringCorrection
deltas = np.zeros(nomega)
weightsMeans = np.zeros(nomega)
weightsM2 = np.zeros(nomega)
weightsErrors = np.zeros(nomega)
for ipath in range(npaths):
    for iomega in range(nomega):
        deltas[iomega] = weights[ipath, iomega] - weightsMeans[iomega]
        weightsMeans[iomega] += deltas[iomega] / (ipath + 1)
        weightsM2[iomega] += deltas[iomega] * (weights[ipath, iomega] - weightsMeans[iomega])
        weightsErrors[iomega] = np.sqrt(weightsM2[iomega] / ipath)
# at this point weightsErrors = [0.19454799 0.20240509 0.20716792 0.20543954 0.19625106]
# roughly consistent with std
# but then there is an additional division by sqrt(nPaths)
for iomega in range(nomega):
    weightsErrors[iomega] = weightsErrors[iomega] / np.sqrt(npaths);
  
print(weightsMeans)  # [0.9997892  0.99953238 0.99211105 1.01015046 1.00047634]
print(weightsErrors) # [0.00615215 0.00640061 0.00655122 0.00649657 0.006206  ]
```
The means agree with `np.mean` but I think the additional division by `sqrt(npaths)` is not needed - before that the values of `weightsErrors` were roughly consistent with `np.std` but there is still a discrepancy.

I think this is because in the first set of loops you divide weightErrors[iomega] by `'ipath`  and I think it should be `npaths`
```
# I propose it should be  this
deltas = np.zeros(nomega)
weightsMeans = np.zeros(nomega)
weightsM2 = np.zeros(nomega)
weightsErrors = np.zeros(nomega)
for ipath in range(npaths):
    for iomega in range(nomega):
        deltas[iomega] = weights[ipath, iomega] - weightsMeans[iomega]
        weightsMeans[iomega] += deltas[iomega] / (ipath + 1)
        weightsM2[iomega] += deltas[iomega] * (weights[ipath, iomega] - weightsMeans[iomega])
        weightsErrors[iomega] = np.sqrt(weightsM2[iomega] / npaths)
print(weightsMeans)  # [0.9997892  0.99953238 0.99211105 1.01015046 1.00047634]
print(weightsErrors) # [0.19445069 0.20230386 0.20706431 0.20533679 0.19615291]
```
Now the results agree with `np.std`
But maybe I've translated it to python wrong...

DannyHindson replied:
I'm not trying to calculate np.std here though. I'm trying to calculate the error on the mean which is np.std/sqrt(n)

RichardWaiteSTFC replied:
Ah the standard error, that's OK then! I still think there will be a small inaccuracy because you divide by `ie` in the loop instead of `nPaths` - actually it is probably better to divide by the number of successful paths (which will have to wait after the loop over nPaths)

DannyHindson replied:
There is a retry in the `ie` loop that means the final value of `ie` will always be `nPaths-1` when the loop completes. As discussed there is in theory a chance the retry mechanism will cause the loop to never complete but I've never seen this happen. I've deliberately divided by `nPaths-1` here because that gives a better estimate (unbiased) of the standard deviation than dividing by `nPaths`.

RichardWaiteSTFC commented on Framework/PythonInterface/mantid/plots/datafunctions.py:
```diff
@@ -924,7 +924,7 @@ def get_errorbar_bounds(container):
     y_segments = _get_y_errorbar_segments(container)
     if y_segments:
-        coords = [array[:, 1] for array in y_segments]
+        coords = [array[:, 1] for array in y_segments if len(array.shape) == 2]
```
Think there is still a bug here if all the errors are `NaN` (which occurs if you run the algorithm with `NeutronPathsMultiple=1`). It produces this error when showing errorbars
```
Traceback (most recent call last):
  File "C:\mantid-conda\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 84, in update_plot_after
    func(*args, **kwargs)
  File "C:\mantid-conda\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 32, in toggle_all_errors
    cls.toggle_error_bars_for(ax, line, make_visible)
  File "C:\mantid-conda\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 73, in toggle_error_bars_for
    cls.update_limits_and_legend(ax, legend_props)
  File "C:\mantid-conda\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 101, in update_limits_and_legend
    ax.relim()
  File "C:\mantid-conda\mantid\Framework\PythonInterface\mantid\plots\mantidaxes.py", line 529, in relim
    min_x, max_x, min_y, max_y = datafunctions.get_errorbar_bounds(container)
  File "C:\mantid-conda\mantid\Framework\PythonInterface\mantid\plots\datafunctions.py", line 928, in get_errorbar_bounds
    max_y = np.max(coords)
  File "<__array_function__ internals>", line 180, in amax
  File "C:\Users\xhg73778\AppData\Local\mambaforge\envs\mantid-developer\lib\site-packages\numpy\core\fromnumeric.py", line 2793, in amax
    return _wrapreduction(a, np.maximum, 'max', axis, None, out,
  File "C:\Users\xhg73778\AppData\Local\mambaforge\envs\mantid-developer\lib\site-packages\numpy\core\fromnumeric.py", line 86, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
ValueError: zero-size array to reduction operation maximum which has no identity
```
Maybe there should be a test for this somewhere as well?

DannyHindson replied:
I've added a test

RichardWaiteSTFC reviewed (changes requested):
Just something to check on the way the variance is calculated (and maybe a loop that could be removed). Also it would be good to fix a bug when getting the data limits from errorbar data if all the errors are NaN

DannyHindson committed:
Use vector of ints instead of double for counts

This was producing a compiler warning on some windows machines

DannyHindson committed:
Fix crash showing error bars when all error are NaN

DannyHindson committed:
Add test for errors containing nan values

Also extend fix to cover xerr as well as yerr

DannyHindson committed:
Make unit test work on matplotlib 3.6

Test with all errors equal to nan caused error in _axes.py due to
modified logic in errorbar function in matplotlib 3.6
Update test to supply masked np array instead of raw nan values because
this is what happens when the UI calls datafunctions.py

RichardWaiteSTFC reviewed (commented):
Thanks for fixing the zero errorbar plotting adding the test. Wen I run the code with 1 neutron path it no longer crashes, but (on the upgraded matplotlib version) when I show error bars on the plot I getting a warning
`Warning: converting a masked element to nan`
I think is related to your change?

DannyHindson commented:
> Thanks for fixing the zero errorbar plotting adding the test. Wen I run the code with 1 neutron path it no longer crashes, but (on the upgraded matplotlib version) when I show error bars on the plot I getting a warning `Warning: converting a masked element to nan` I think is related to your change?

I think I've fixed a couple of crashes relating to errors being equal to nan and in their place there is now this warning that's due to an unrelated choice to plot masked numpy arrays in matplotlib. This choice is made in the function `get_spectrum` in datafunctions.py where it converts the y and dy arrays to masked arrays before plotting the data. This is from ~L344:
```
    y = np.ma.masked_invalid(y)
    if dy is not None:
        dy = np.ma.masked_invalid(dy)
```
During the process of plotting the data, matplotlib calls numpy to convert the arrays back to standard arrays which involves converting any masked elements back to NaN and it generates the warning. It's a side effect of plotting masked arrays I think.

So I'm thinking it's probably OK? Unless there's some way of suppressing the warning? Not passing masked arrays out of `get_spectrum` might create other problems. I'm assuming someone did this to avoid nan values causing trouble elsewhere.

Re the other point about the double scattering profiles looking odd. I've just checked the results on main and they're the same. I'll perhaps have a look at this anyway (separately) but I don't think it's related to the changes in this PR.

RichardWaiteSTFC reviewed (approved):
Thank for looking into the warning, I'm happy to approve without supressing the warning as it all functions correctly and major changes to the masking/handling of NaNs is more likely to introduce a bug quite close to the release! Agreed that the curve for 1 double scattering path is a separate issue.
